### Importing Libraries 

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

### Step 1 Importing Dataset

In [2]:
combined_dataset = pd.read_pickle(r"C:\Users\Windows\Instacart Basket Analysis\02 Data\Prepared Data\combined_dataset.pkl")

In [3]:
combined_dataset.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 31 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   product_id              int32   
 1   product_name            category
 2   aisle_id                int16   
 3   department_id           int8    
 4   prices                  float64 
 5   order_id                int32   
 6   user_id                 int32   
 7   order_number            int8    
 8   orders_day_of_week      int8    
 9   hour_of_day_ordered     int8    
 10  days_since_prior_order  float16 
 11  add_to_cart_order       int16   
 12  reordered               int8    
 13  busiest_days            category
 14  busiest_period_of_day   category
 15  price_range_loc         category
 16  max_order               int8    
 17  loyalty_flag            category
 18  avg_spending            float64 
 19  spender_flag            category
 20  last_order              float16 
 21  ord_fr

### Step 2 - The personal identifying information (PII) includes first name and surname. The information can be traced back to the user and once the user is known, further information about that person could well be exposed such as their credit card information, social security number, etc. As a junior analyst, it is wise to seek the company's policy with regard to dealing with PII and consult with colleagues to clarify.¶

### Step 3 - Creating the "Region" column based on the "State" column

In [32]:
combined_dataset.shape

(32404859, 33)

In [33]:
combined_dataset.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        19             19     5.8   
1           1  Chocolate Sandwich Cookies        19             19     5.8   
2         907        Premium Sliced Bacon        12             12    20.0   
3         907        Premium Sliced Bacon        12             12    20.0   
4        1000                    Apricots        10             10    12.9   

   order_id  user_id  order_number  orders_day_of_week  hour_of_day_ordered  \
0   3139998      138            28                   6                   11   
1   1977647      138            30                   6                   17   
2   3160996      138             1                   5                   13   
3   2254091      138            10                   5                   14   
4    505689      138             9                   6                   12   

   ...  Surname  Gender      State Age date_joined n_dependants  fam_status  \
0  ...      Cox    Male  Minnesota  81    8/1/2019            1     married   
1  ...      Cox    Male  Minnesota  81    8/1/2019            1     married   
2  ...      Cox    Male  Minnesota  81    8/1/2019            1     married   
3  ...      Cox    Male  Minnesota  81    8/1/2019            1     married   
4  ...      Cox    Male  Minnesota  81    8/1/2019            1     married   

  income   Region exclusion  
0  49620  Midwest        No  
1  49620  Midwest        No  
2  49620  Midwest        No  
3  49620  Midwest        No  
4  49620  Midwest        No  

[5 rows x 33 columns]

In [29]:
# col_1 = []

# for value in combined_dataset["State"]:
#   if value in ['Maine','New Hampshire','Vermont','Massachusetts', 'Rhode Island', 'Connecticut', 'New York','Pennsylvania',
#                'New Jersey']:
#     col_1.append("Northeast")
#   elif value in ['Wisconsin', 'Michigan','Illinois','Indiana','Ohio','North Dakota','South Dakota','Nebraska',
#                  'Kansas','Minnesota','Iowa','Missouri']:
#     col_1.append("Midwest")
#   elif value in ['Delaware','Maryland','District of Columbia','Virginia','West Virginia','North Carolina','South Carolina',
#                 'Georgia','Florida','Kentucky','Tennessee','Mississippi','Alabama','Oklahoma','Texas','Arkansas','Louisiana']:
#     col_1.append("South")
#   elif value in ['Idaho','Montana','Wyoming','Nevada','Utah','Colorado','Arizona','New Mexico','Alaska','Washington''Oregon',
#                 'California','Hawaii']:
#     col_1.append('West')
  
 


In [34]:
# create lists of regions 

Northeast=['Maine','New Hampshire','Vermont','Massachusetts','Rhode Island','Connecticut','New York','Pennsylvania','New Jersey']
Midwest=['Wisconsin','Michigan','Illinois','Indiana','Ohio','North Dakota','South Dakota','Nebraska','Kansas','Minnesota','Iowa','Missouri']
South=['Delaware','Maryland','District of Columbia','Virginia','West Virginia','North Carolina','South Carolina','Georgia','Florida','Kentucky', 'Tennessee','Mississippi','Alabama','Oklahoma','Texas','Arkansas','Louisiana']
West=['Idaho','Montana','Wyoming','Nevada','Utah','Colorado','Arizona','New Mexico','Alaska','Washington','Oregon','California','Hawaii']

In [36]:
# Creating regional flag column

combined_dataset.loc[combined_dataset['State'].isin(Northeast),'Region']='Northeast'
combined_dataset.loc[combined_dataset['State'].isin(Midwest),'Region']='Midwest'
combined_dataset.loc[combined_dataset['State'].isin(South),'Region']='South'
combined_dataset.loc[combined_dataset['State'].isin(West),'Region']='West'

In [37]:
len(combined_dataset['Region'])

32404859

In [38]:
combined_dataset['Region'].value_counts(dropna=False)

South        10791885
West          8292913
Midwest       7597325
Northeast     5722736
NaN                 0
Name: Region, dtype: int64

In [39]:
combined_dataset['Region']=combined_dataset['Region'].astype('category')

In [40]:
combined_dataset.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 33 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   product_id              int32   
 1   product_name            category
 2   aisle_id                int16   
 3   department_id           int8    
 4   prices                  float64 
 5   order_id                int32   
 6   user_id                 int32   
 7   order_number            int8    
 8   orders_day_of_week      int8    
 9   hour_of_day_ordered     int8    
 10  days_since_prior_order  float16 
 11  add_to_cart_order       int16   
 12  reordered               int8    
 13  busiest_days            category
 14  busiest_period_of_day   category
 15  price_range_loc         category
 16  max_order               int8    
 17  loyalty_flag            category
 18  avg_spending            float64 
 19  spender_flag            category
 20  last_order              float16 
 21  ord_fr

In [11]:
# checking for difference in spending habits between different US regions

crosstab = pd.crosstab(combined_dataset['spender_flag'], combined_dataset['Region'],normalize=True, margins=True)

In [12]:
crosstab

Region         Midwest       NaN  Northeast     South      West       All
spender_flag                                                             
High spender  0.004813  0.000857   0.003340  0.006471  0.004091  0.019573
Low spender   0.229637  0.038312   0.173261  0.326562  0.212655  0.980427
All           0.234450  0.039169   0.176601  0.333033  0.216746  1.000000

#### As we can see, 33.3% of the spenders are from the South, who make up the majority of the consumers followed by Midwest, West and Northeast.   98% of the spenders are considered "Low spenders".  In terms of "High spenders", the South is also in first place, followed by Midwest, West, and Northeast.

### Step  4

In [13]:
# Flag low-activity customers with less than 5 orders 

combined_dataset.loc[combined_dataset['max_order'] < 5, 'exclusion'] = 'Yes'

In [14]:
combined_dataset.loc[combined_dataset['max_order'] > 5, 'exclusion'] = 'No'

In [15]:
combined_dataset.head(5)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        19             19     5.8   
1           1  Chocolate Sandwich Cookies        19             19     5.8   
2         907        Premium Sliced Bacon        12             12    20.0   
3         907        Premium Sliced Bacon        12             12    20.0   
4        1000                    Apricots        10             10    12.9   

   order_id  user_id  order_number  orders_day_of_week  hour_of_day_ordered  \
0   3139998      138            28                   6                   11   
1   1977647      138            30                   6                   17   
2   3160996      138             1                   5                   13   
3   2254091      138            10                   5                   14   
4    505689      138             9                   6                   12   

   ...  Surname  Gender      State Age date_joined n_dependants  fam_status  \
0  ...      Cox    Male  Minnesota  81    8/1/2019            1     married   
1  ...      Cox    Male  Minnesota  81    8/1/2019            1     married   
2  ...      Cox    Male  Minnesota  81    8/1/2019            1     married   
3  ...      Cox    Male  Minnesota  81    8/1/2019            1     married   
4  ...      Cox    Male  Minnesota  81    8/1/2019            1     married   

  income   Region exclusion  
0  49620  Midwest        No  
1  49620  Midwest        No  
2  49620  Midwest        No  
3  49620  Midwest        No  
4  49620  Midwest        No  

[5 rows x 33 columns]

In [16]:
excluded_customers = combined_dataset[combined_dataset['exclusion'] == 'Yes']

In [17]:
excluded_customers.head(5)

product_id                            product_name  aisle_id  \
340           1              Chocolate Sandwich Cookies        19   
341           1              Chocolate Sandwich Cookies        19   
342        3260  Chips Ahoy!/Nutter Butter/Oreo Cookies        19   
343        3260  Chips Ahoy!/Nutter Butter/Oreo Cookies        19   
344        3260  Chips Ahoy!/Nutter Butter/Oreo Cookies        19   

     department_id  prices  order_id  user_id  order_number  \
340             19     5.8    652770      764             1   
341             19     5.8   1813452      764             3   
342             19     1.7    652770      764             1   
343             19     1.7    705212      764             2   
344             19     1.7   1813452      764             3   

     orders_day_of_week  hour_of_day_ordered  ...  Surname  Gender      State  \
340                   3                   13  ...    Myers  Female  Wisconsin   
341                   4                   17  ...    Myers  Female  Wisconsin   
342                   3                   13  ...    Myers  Female  Wisconsin   
343                   2                   10  ...    Myers  Female  Wisconsin   
344                   4                   17  ...    Myers  Female  Wisconsin   

    Age date_joined n_dependants  fam_status income   Region exclusion  
340  40    2/9/2020            3     married  31308  Midwest       Yes  
341  40    2/9/2020            3     married  31308  Midwest       Yes  
342  40    2/9/2020            3     married  31308  Midwest       Yes  
343  40    2/9/2020            3     married  31308  Midwest       Yes  
344  40    2/9/2020            3     married  31308  Midwest       Yes  

[5 rows x 33 columns]

In [18]:
path = r"C:\Users\Windows\Instacart Basket Analysis"

In [19]:
excluded_customers.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'excluded_customers.pkl'))

In [20]:
# create a dataset with only customers with more than 5 orders
df = combined_dataset[combined_dataset['exclusion'] == 'No']

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30171424 entries, 0 to 32404847
Data columns (total 33 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   product_id              int32   
 1   product_name            category
 2   aisle_id                int16   
 3   department_id           int8    
 4   prices                  float64 
 5   order_id                int32   
 6   user_id                 int32   
 7   order_number            int8    
 8   orders_day_of_week      int8    
 9   hour_of_day_ordered     int8    
 10  days_since_prior_order  float16 
 11  add_to_cart_order       int16   
 12  reordered               int8    
 13  busiest_days            category
 14  busiest_period_of_day   category
 15  price_range_loc         category
 16  max_order               int8    
 17  loyalty_flag            category
 18  avg_spending            float64 
 19  spender_flag            category
 20  last_order              float16 
 21  ord_fr

In [22]:
df['exclusion'] = df['exclusion'].astype('category')

C:\Users\Windows\AppData\Local\Temp\ipykernel_8736\1162234098.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['exclusion'] = df['exclusion'].astype('category')


In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30171424 entries, 0 to 32404847
Data columns (total 33 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   product_id              int32   
 1   product_name            category
 2   aisle_id                int16   
 3   department_id           int8    
 4   prices                  float64 
 5   order_id                int32   
 6   user_id                 int32   
 7   order_number            int8    
 8   orders_day_of_week      int8    
 9   hour_of_day_ordered     int8    
 10  days_since_prior_order  float16 
 11  add_to_cart_order       int16   
 12  reordered               int8    
 13  busiest_days            category
 14  busiest_period_of_day   category
 15  price_range_loc         category
 16  max_order               int8    
 17  loyalty_flag            category
 18  avg_spending            float64 
 19  spender_flag            category
 20  last_order              float16 
 21  ord_fr

In [24]:
# dataset 
df.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'updated_dataset.pkl'))

In [25]:
df[['department_id','aisle_id','product_name']].head()

department_id  aisle_id                product_name
0             19        19  Chocolate Sandwich Cookies
1             19        19  Chocolate Sandwich Cookies
2             12        12        Premium Sliced Bacon
3             12        12        Premium Sliced Bacon
4             10        10                    Apricots

In [26]:
df['product_name'].value_counts(dropna=False)

Banana                                   442488
Bag of Organic Bananas                   359163
Organic Strawberries                     251133
Organic Baby Spinach                     226320
Organic Hass Avocado                     203216
                                          ...  
Original Jerky                                0
Natural Peanut Butter With Flaxseed           0
Forest Body Wash                              0
String Of Pearl White Sprinkles               0
Hydrating Moisturizing Lotion for Men         0
Name: product_name, Length: 49661, dtype: int64

In [27]:
# Profile A: Low-income individuals with dependents.
A = (df['income'] < 50000) & (df['n_dependants'] > 0)
df.loc[A, 'profile'] = 'Low_Income_with_Dependants'

C:\Users\Windows\AppData\Local\Temp\ipykernel_8736\1954271384.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[A, 'profile'] = 'Low_Income_with_Dependants'


In [28]:
df[df['profile'] == 'Low_Income_with_Dependants']

product_id                product_name  aisle_id  department_id  \
0                  1  Chocolate Sandwich Cookies        19             19   
1                  1  Chocolate Sandwich Cookies        19             19   
2                907        Premium Sliced Bacon        12             12   
3                907        Premium Sliced Bacon        12             12   
4               1000                    Apricots        10             10   
...              ...                         ...       ...            ...   
32404843       49235         Organic Half & Half        16             16   
32404844       49235         Organic Half & Half        16             16   
32404845       49235         Organic Half & Half        16             16   
32404846       49235         Organic Half & Half        16             16   
32404847       49235         Organic Half & Half        16             16   

          prices  order_id  user_id  order_number  orders_day_of_week  \
0            5.8   3139998      138            28                   6   
1            5.8   1977647      138            30                   6   
2           20.0   3160996      138             1                   5   
3           20.0   2254091      138            10                   5   
4           12.9    505689      138             9                   6   
...          ...       ...      ...           ...                 ...   
32404843     1.8   3310701    27382             5                   3   
32404844     1.8    527883    27382             6                   1   
32404845     1.8    685496    27382             7                   4   
32404846     1.8   1224680    27382             8                   1   
32404847     1.8   2223687    27382             9                   2   

          hour_of_day_ordered  ...  Gender      State  Age date_joined  \
0                          11  ...    Male  Minnesota   81    8/1/2019   
1                          17  ...    Male  Minnesota   81    8/1/2019   
2                          13  ...    Male  Minnesota   81    8/1/2019   
3                          14  ...    Male  Minnesota   81    8/1/2019   
4                          12  ...    Male  Minnesota   81    8/1/2019   
...                       ...  ...     ...        ...  ...         ...   
32404843                    9  ...  Female    Vermont   68  11/25/2019   
32404844                   12  ...  Female    Vermont   68  11/25/2019   
32404845                    8  ...  Female    Vermont   68  11/25/2019   
32404846                    7  ...  Female    Vermont   68  11/25/2019   
32404847                   12  ...  Female    Vermont   68  11/25/2019   

         n_dependants fam_status  income     Region  exclusion  \
0                   1    married   49620    Midwest         No   
1                   1    married   49620    Midwest         No   
2                   1    married   49620    Midwest         No   
3                   1    married   49620    Midwest         No   
4                   1    married   49620    Midwest         No   
...               ...        ...     ...        ...        ...   
32404843            3    married   37867  Northeast         No   
32404844            3    married   37867  Northeast         No   
32404845            3    married   37867  Northeast         No   
32404846            3    married   37867  Northeast         No   
32404847            3    married   37867  Northeast         No   

                             profile  
0         Low_Income_with_Dependants  
1         Low_Income_with_Dependants  
2         Low_Income_with_Dependants  
3         Low_Income_with_Dependants  
4         Low_Income_with_Dependants  
...                              ...  
32404843  Low_Income_with_Dependants  
32404844  Low_Income_with_Dependants  
32404845  Low_Income_with_Dependants  
32404846  Low_Income_with_Dependants  
32404847  Low_Income_with_Dependants  

[2430109 rows x 34 columns]